payment-request/PaymentRequestUpdateEvent/updateWith-call-immediate-manual.https.html

```diff
@@ -0,0 +1,171 @@
+      request.addEventListener(
+        "shippingaddresschange",
+        async ev => {
+          // spin the event loop, sets [[waitForUpdate]] to true.
+          await Promise.resolve();
+          try {
+            ev.updateWith(validDetails);
+            resolve(); // This is bad.
+          } catch (err) {
+            reject(err); // this is good.
+          }
+        },
+        { once: true }
+      );
+    });
+    const response = await request.show();
+    await promise_rejects(
+      t,
+      "InvalidStateError",
+      shippingOptionEventPromise,
+      "The event loop already spun, so [[waitForUpdate]] is now true"
+    );
+    await response.complete();
+  }, testName.trim());
+}
+
+function testSubsequentUpdateWithCalls({ textContent: testName }) {
+  promise_test(async t => {
+    const request = new PaymentRequest(
+      validMethods,
+      validDetails,
+      validOptions
+    );
+    const eventPromise = new Promise((resolve, reject) => {
+      request.addEventListener("shippingaddresschange", async ev => {
+        const p = Promise.resolve(validDetails);
+        ev.updateWith(p);
+        await p;
+        try {
+          ev.updateWith(validDetails);
+          resolve(); // this is bad, we should never get to here.
+        } catch (err) {
+          reject(err); // this is good!
+        }
+      });
+    });
+    const responsePromise = request.show();
+    await promise_rejects(t, "InvalidStateError", eventPromise);
+    const response = await responsePromise;
+    await response.complete();
+  }, testName.trim());
+}
+
+function testRecycleEvents({ textContent: testName }) {
+  promise_test(async t => {
+    const request = new PaymentRequest(
+      validMethods,
+      validDetails,
+      validOptions
+    );
+    const addressChangedPromise = new Promise(resolve => {
+      request.addEventListener("shippingaddresschange", resolve, {
+        once: true,
+      });
+    });
+    const optionChangedPromise = new Promise(resolve => {
+      request.addEventListener("shippingoptionchange", resolve, { once: true });
+    });
+
+    const responsePromise = request.show();
+
+    // Let's wait for the address to change.
+    const addressChangeEvent = await addressChangedPromise;
+    addressChangeEvent.updateWith(validDetails);
+
+    // Let's wait for the shippingOption
+    const optionChangeEvent = await optionChangedPromise;
+
+    // The addressChangeEvent was already consumed and expired.
+    assert_throws(
+      new TypeError(),
+      () => {
+        addressChangeEvent.updateWith(validDetails);
+      },
+      "addressChangeEvent [[waitingForUpdate]] is true, so it must throw"
+    );
+
+    // But optionChangeEvent is still usable tho, so...
+    optionChangeEvent.updateWith(validDetails);
+
+    assert_throws(
+      "InvalidStateError",
+      () => {
+        optionChangeEvent.updateWith(validDetails);
+      },
+      "optionChangeEvent [[waitingForUpdate]] is true, so it must throw"
```

Comment:
Here is a rase condition. We can't not sure which one is executed between the first optionChangeEvent.updateWith() resolve callback and the second optionChangeEvent.updateWith(). If the resolve callback is executed first, optionChangeEvent.[[waitingForUpdate]] will become false.
If we want to make sure [[waitingForUpdate]] is true, using a pending promise in the first optionChangeEvent.updateWith() probably is a good solution. And we can add two assertions before and after resolving the pending promise. The before one should get InvalidStateError, since the optionChangeEvent.[waitingForUpdate] is true. And the second should throw an error, maybe InvalidStateError, since the event is expired. 

By the way, in fact, the spec doesn't mention what happens if the merchant calls PaymentRequestUpdateEvent.updateWith repeatlly. In my opinion, for each PaymentRequestUpdateEvent, updateWith can only be called one time.

Reply:
> Here is a rase condition. We can't not sure which one is executed between the first optionChangeEvent.updateWith() resolve callback and the second optionChangeEvent.updateWith()

In the spec, .updateWith() immediately sets `[waitingForUpdate]` to true (before we go async):

> 8. Set event.[[waitForUpdate]] to true.

And step 3 guards:

> 3. If event.[[waitForUpdate]] is true, then throw an "InvalidStateError" DOMException.

Reply:
> By the way, in fact, the spec doesn't mention what happens if the merchant calls PaymentRequestUpdateEvent.updateWith repeatlly. In my opinion, for each PaymentRequestUpdateEvent, updateWith can only be called one time. 

The:

> 3. If event.[[waitForUpdate]] is true, then throw an "InvalidStateError" DOMException.

Prevents calling .updateWith() multiple times.